### PR TITLE
Settings Page: update

### PR DESCRIPTION
**Pull request by ole-na**
Settings form on the Settings Page: 
- add route fields and distance slider,
- change warning category,
- validate entries ("start" and "end" are required, points are optional, at least one warning category should be selected)
- "start" and "end" are always available, but other points can be dynamically added
- on reset (click on the reset button): only "start" and "end" should be available with empty values, all other point fields should be removed
- on click on trashcan of "start"/"end": set its values to empty
- on click on trashcan of other points: delete this point field

### CODE REVIEW

File: frontend/src/components/settings/route/RouteSetting.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        setRoute({start: "", end: "", points: []})
+        setErrors({start: false, end: false, points: []})
+        setPoints([])
+    }
+
+    /* useEffect(() => {
+        console.log("R:", route, "E:", errors, "P", points)
+    }, [route, errors, points]) */
+
+    // update hooks for route, errors and points: index of end ist
```
Comment (ole-na) [inline frontend/src/components/settings/route/RouteSetting.js:64]: Refactoring: use function handleChange(event) for start and end and handleChangePoint(event, index) 

const handleChange = (event) => {
        const value = event.target.value
        const errorValue = (value === "" || !routeServices.checkRoutePointValue(value))
        setRoute({...route, [item]: value});
        setErrors({...errors, [item]: errorValue});
}
const handleChangePoint = (event, index) => {
        const value = event.target.value
        const errorValue = (value === "" || !routeServices.checkRoutePointValue(value))

        const newRoutePoints = route.points
        newRoutePoints[index] = value
        setRoute({...route, [item]: newRoutePoints});

        const newErrors = errors.points
        newErrors[index] = errorValue;
        setErrors({...errors, [item]: newErrors});
    }

File: frontend/src/components/settings/route/RouteSetting.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                                const key = "pointField" + point
+                                return (
+                                    <Grid key={key} item xs={12} className="route-point-field-container">
+                                        <RoutePoint name={key}
+                                                    label="Point"
+                                                    required={false}
+                                                    value={route.points[index]}
+                                                    error={errors.points[index]}
+                                                    iconColor={grey}
+                                                    onChange={(event) => handleChange(event, "points", index)}
```
Comment (ole-na) [inline frontend/src/components/settings/route/RouteSetting.js:155]: Refactoring: `onChange={(event) => handleChangePoint(event, index)}`

File: frontend/src/pages/SettingsPage.js [hunk 3]
```diff
@@ -1,19 +1,57 @@
-import React from "react";
-import Button from "@material-ui/core/Button";
-import TypeWarningSetting from "../components/settings/TypeWarningSetting";
+import React, {useState} from "react";
+// import { getSettings } from '../services/settingsApi'
```
Comment (jamarob) [inline frontend/src/pages/SettingsPage.js:2]:
```suggestion

```

File: frontend/src/pages/SettingsPage.js [hunk 4]
```diff
@@ -1,19 +1,57 @@
...
+            start: "",
+            end: "",
+            points: []
+        }
+    })
+
+    const [isLoading, setIsLoading] = useState(false);
+    const [error, setError] = useState(false);
+
+    const fetchData = async () => {
```
Comment (jamarob) [inline frontend/src/pages/SettingsPage.js:22]: When is this function called, I think you forgot to write a useEffect for it? It doesn't do anything atm (requst is commented out). Try to include only code that works in your pr.

File: frontend/src/pages/SettingsPage.js [hunk 5]
```diff
@@ -1,19 +1,57 @@
...
-                <TypeWarningSetting />
-                <Button variant="contained" color="primary" type="submit" >Save</Button>
-            </form>
+
+            {isLoading && <p>Loading...</p>}
+
+            {!isLoading && error ? (
+                <p>Oh no something went wrong!</p>
+
+            ) : null}
```
Comment (jamarob) [inline frontend/src/pages/SettingsPage.js:49]:
```suggestion
            {!isLoading && error && <p>Oh no something went wrong!</p>}
```

No need for a ternary here

File: frontend/src/pages/SettingsPage.js [hunk 6]
```diff
@@ -1,19 +1,57 @@
...
+
+            {settings ? <SettingsForm settings={settings} /> : null}
+
+            {!settings && !isLoading ? <div>No settings data yet</div> : null}
```
Comment (jamarob) [inline frontend/src/pages/SettingsPage.js:53]: see comment above

**Review by jamarob — COMMENTED** [review]
Good job! Left some comments.

Try to do a little bit more cleanup before opening a pr, there are a some comments and commented out code. Please delete these, a better way to document your code is by introducing functions with descriptive names.

File: frontend/src/components/settings/DistanceSetting.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        value: 15,
+        label: '15',
+    },
+    {
+        value: 30,
+        label: '30',
+    },
+    {
+        value: 50,
+        label: '50nm',
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/DistanceSetting.js:24]: It's puffer. The warning area should be max 15 nm

File: frontend/src/components/settings/DistanceSetting.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+export default function DistanceSetting(props) {
+    const classes = useStyles();
+    const [distance, setDistance] = useState(props.distance)
+
+    const handleChange = (event) => {
+        setDistance(event.target.value);
+    };
+
+    return (
+        <div className={props.settingClasses.root}>
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/DistanceSetting.js:49]: good idea, i selected the difficult and not optimal way

File: frontend/src/components/settings/route/RouteSetting.js [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    // initial array hook with input name values for points (not start and end)
+    const pointsValues = []
+    for (let i in props.route.points) {
+        pointsValues[i] = "point" + i
+    }
+    const [points, setPoints] = useState(pointsValues)
+
+    // check required values for start and end only, other points are optional
+    const { start, end } = route;
+    const error = [ start, end ].filter((v) => v).length === 0;
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/route/RouteSetting.js:50]: ok

File: frontend/src/components/settings/route/RouteSetting.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        newRoutePoints.push("")
+        setRoute({...route, points: newRoutePoints});
+
+        // add new point to errors.points
+        const newErrorPoints = errors.points
+        newErrorPoints.push(false)
+        setErrors({...errors, points: newErrorPoints});
+
+        // find the biggest number in points array and use this value for the new element (+1)
+        const pointName = points.length > 0 ? (Math.max.apply(null, points) + 1) : 1
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/route/RouteSetting.js:100]: yes, it works too

File: frontend/src/components/settings/route/RouteSetting.js [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+        // update errors: delete current error from error.points list
+        const newErrors = errors.points
+        newErrors.splice(index, 1)
+        setErrors({...errors, points: newErrors});
+
+        // update points:  delete current error from points list
+        const newPoints = points
+        newPoints.splice(index, 1)
+        setPoints(newPoints)
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/route/RouteSetting.js:123]: ok, I have changed to copy "[...points]" instead "points". Also for start, end in another functions

File: frontend/src/pages/SettingsPage.js (hunk 5 above)
Comment (ole-na) [reply frontend/src/pages/SettingsPage.js]: ok

File: frontend/src/components/settings/route/RouteSetting.js [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    /* useEffect(() => {
+        console.log("R:", route, "E:", errors, "P", points)
+    }, [route, errors, points]) */
+
+    // update hooks for route, errors and points: index of end ist
+    const handleChange = (event, item, index) => {
+        const value = event.target.value
+        const errorValue = (value === "" || !routeServices.checkRoutePointValue(value))
```
[the remark this answers is not in the document]
Comment (ole-na) [reply frontend/src/components/settings/route/RouteSetting.js:76]: you are right, the check for empty string is already in regexp

File: frontend/src/pages/SettingsPage.js (hunk 4 above)
Comment (ole-na) [reply frontend/src/pages/SettingsPage.js]: ok, I have added useeffect. But I leave this commented out code inside, I have to turn it up at the end of the week anyway

File: frontend/src/pages/SettingsPage.js (hunk 3 above)
Comment (ole-na) [reply frontend/src/pages/SettingsPage.js:2]: see the next comment

**Pull request merged**